cmd/promtool/sd.go

```diff
@@ -0,0 +1,143 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/go-kit/log"
+
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/discovery"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/pkg/relabel"
+)
+
+type sdCheckResult struct {
+	Target          string
+	OriginalLabels  labels.Labels
+	AfterRelabeling labels.Labels
+}
+
+// CheckSD performs service discovery for the given job name and reports the results.
+func CheckSD(sdConfigFiles, sdJobName string, sdTimeout time.Duration) int {
+	logger := log.NewLogfmtLogger(log.NewSyncWriter(os.Stderr))
+
+	cfg, err := config.LoadFile(sdConfigFiles, false, logger)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, "Cannot load config", err)
+		return 2
+	}
+
+	var discoveryConfigs discovery.Configs
+	var relabelConfigs []*relabel.Config
+	jobs := []string{}
+	jobMatched := false
+	for _, v := range cfg.ScrapeConfigs {
+		jobs = append(jobs, v.JobName)
+		if v.JobName == sdJobName {
+			jobMatched = true
+			discoveryConfigs = v.ServiceDiscoveryConfigs
+			relabelConfigs = v.RelabelConfigs
+			break
+		}
+	}
+
+	if !jobMatched {
+		fmt.Fprintf(os.Stderr, "Job %s not found. Select one of:\n", sdJobName)
+		for _, job := range jobs {
+			fmt.Fprintf(os.Stderr, "\t%s\n", job)
+		}
+		return 1
+	}
+
+	targetGroupChan := make(chan []*targetgroup.Group)
+	ctx, cancel := context.WithTimeout(context.Background(), sdTimeout)
+	defer cancel()
+
+	for _, cfg := range discoveryConfigs {
+		d, err := cfg.NewDiscoverer(discovery.DiscovererOptions{Logger: logger})
+		if err != nil {
+			fmt.Fprintln(os.Stderr, "Could not create new discoverer", err)
+			return 2
+		}
+		go d.Run(ctx, targetGroupChan)
+	}
+
+	var targetGroups []*targetgroup.Group
+	sdCheckResults := make(map[string][]*targetgroup.Group)
+	returnCode := 0
+	loop := true
+	for loop {
+		select {
+		case targetGroups = <-targetGroupChan:
+			for _, tg := range targetGroups {
+				sdCheckResults[tg.Source] = append(sdCheckResults[tg.Source], tg)
+			}
+		case <-ctx.Done():
+			loop = false
+			break
+		}
+	}
+
+	for _, tgs := range sdCheckResults {
+		for _, result := range getSDCheckResult(tgs, relabelConfigs) {
+			res, err := json.MarshalIndent(result, "", "\t")
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "Could not marshal result json: %s\n", err)
+				returnCode = 2
+			}
+
+			fmt.Println(string(res))
```

Comment:
This isn't outputting valid JSON, as it's a bunch of objects concatenated together.

I think this might be a case where it makes sense to just output the wrapping array ourselves, as that way we can keep the streaming output for people running it in a terminal, but it will parse as JSON too.

I played with it, something like:

```diff
--- a/cmd/promtool/sd.go
+++ b/cmd/promtool/sd.go
@@ -96,18 +96,31 @@ func CheckSD(sdConfigFiles, sdJobName string, sdTimeout time.Duration) int {
                }
        }

+       i := 0
+
+       fmt.Println("[")
+
        for _, tgs := range sdCheckResults {
                for _, result := range getSDCheckResult(tgs, relabelConfigs) {
-                       res, err := json.MarshalIndent(result, "", "\t")
+                       res, err := json.MarshalIndent(result, "  ", "  ")
                        if err != nil {
                                fmt.Fprintf(os.Stderr, "Could not marshal result json: %s\n", err)
                                returnCode = 2
                        }

-                       fmt.Println(string(res))
+                       if i > 0 {
+                               fmt.Println(",")
+                       }
+                       fmt.Print("  ", string(res))
+                       i++
                }
        }

+       if i > 0 {
+               fmt.Println("")
+       }
+       fmt.Println("]")
+
        return returnCode
 }
```

(I adjusted the tabs with spaces, as that looks nicer IMO)

This then works when I do:
```
./promtool check sd ~/.../prometheus.yml blackbox-http 1s | jq .
```